File: Data Structures & Algorithms/eating-bananas/submission-4.java
public class Solution {
    public int minEatingSpeed(int[] piles, int h) {

        // merge sort 
        // mid = left + (right - left) / 2
        // while (left < right )
        // left = mid + 1
        // right = mid - 1
        int eating_rate = 0;
        int hours = piles.length;

        for (int i = 0; i < piles.length; i++) {
            if (piles[i] > eating_rate) {
                eating_rate = piles[i];
            }
        }

        if (eating_rate == 0) {
            return 0;
        }

        System.out.println(eating_rate);

        int left = 1;
        int right = eating_rate;

        while (left < right) {
            int mock_eating_rate = left + (right - left) / 2;
            System.out.println(mock_eating_rate);
            int count_hours = 0;
            
            for (int i = 0; i < piles.length; i++) {
                count_hours += (piles[i] / mock_eating_rate);
                if (piles[i] % mock_eating_rate != 0) {
                    count_hours ++;
                }
            }

            if (count_hours <= h) {
                right = mock_eating_rate;
                eating_rate = mock_eating_rate; 
            } else {
                left = mock_eating_rate + 1;
            }

        }

        return eating_rate;
    }
}

File: Data Structures & Algorithms/search-2d-matrix/submission-4.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        r1 = 0
        r2 = len(matrix) - 1
        mid = 0
        while r1 <= r2: 
            mid = (r1 + r2) // 2 
            if target == matrix[mid][0] or target == matrix[mid][-1]:
                return True
            elif target > matrix[mid][0] and target < matrix[mid][-1]:
                break
            elif target < matrix[mid][0]:
                r2 = mid - 1
            else:
                r1 = mid + 1
        
        l1, l2 = 0, len(matrix[0]) - 1
        while l1 <= l2:
            m = (l1 + l2) // 2
            if target == matrix[mid][m]: 
                return True
            elif target < matrix[mid][m]:
                l2 = m - 1
            else: 
                l1 = m + 1
        
        return False



